docs/configuration.md
# Configuration

Caluma is a [12factor app](https://12factor.net/) which means that configuration is stored in environment variables.
Different environment variable types are explained at [django-environ](https://github.com/joke2k/django-environ#supported-types).

## Common

A list of configuration options which you might need to configure to get Caluma started in your environment.

* `UID`: The user ID used in the container
* `SECRET_KEY`: A secret key used for cryptography. This needs to be a random string of a certain length. See [more](https://docs.djangoproject.com/en/2.1/ref/settings/#std:setting-SECRET_KEY).
* `ALLOWED_HOSTS`: A list of hosts/domains your service will be served from. See [more](https://docs.djangoproject.com/en/2.1/ref/settings/#allowed-hosts).
* `DATABASE_HOST`: Host to use when connecting to database (default: localhost)
* `DATABASE_PORT`: Port to use when connecting to database (default: 5432)
* `DATABASE_NAME`: Name of database to use (default: caluma)
* `DATABASE_USER`: Username to use when connecting to the database (default: caluma)
* `DATABASE_PASSWORD`: Password to use when connecting to database
* `LANGUAGE_CODE`: Default language defined as fallback (default: en)
* `LANGUAGES`: List of supported language codes (default: all available)
* `LOG_LEVEL`: [Log level](https://docs.djangoproject.com/en/1.11/topics/logging/#loggers) of messages to write to output (default: INFO)

## Authentication and authorization

If you want to connect to Caluma you need an
[IAM](https://en.wikipedia.org/wiki/Identity_management) supporting
OpenID Connect. If you don't have this available in your environment already,
you might want to consider using [Keycloak](https://www.keycloak.org/).

Caluma expects a bearer token to be passed on as [Authorization Request Header Field](https://tools.ietf.org/html/rfc6750#section-2.1)

* `OIDC_USERINFO_ENDPOINT`: Url of userinfo endpoint as [described](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)
* `OIDC_GROUPS_CLAIM`: Name of claim to be used to represent groups (default: caluma_groups)
* `OIDC_USERNAME_CLAIM`: Name of claim to be used to represent the username (default: sub)
* `OIDC_BEARER_TOKEN_REVALIDATION_TIME`: Time in seconds before bearer token validity is verified again. For best security token is validated on each request per default. It might be helpful though in case of slow Open ID Connect provider to cache it. It uses [cache](#cache) mechanism for memorizing userinfo result. Number has to be lower than access token expiration time. (default: 0)
* `CALUMA_OIDC_USER_FACTORY`: User object factory (default: `caluma.caluma_user.models.OIDCUser`). Use it to provide a custom OIDC user object. The factory is expected to accept a mandatory `token` parameter and two optional parameters `userinfo` and `introspection`. Only one of them will be filled, depending on which OIDC endpoint the user information comes from.

## Cache

* `CACHE_BACKEND`: [cache backend](https://docs.djangoproject.com/en/1.11/ref/settings/#backend) to use (default: django.core.cache.backends.locmem.LocMemCache)
* `CACHE_LOCATION`: [location](https://docs.djangoproject.com/en/1.11/ref/settings/#std:setting-CACHES-LOCATION) of cache to use

## CORS headers

Per default no CORS headers are set but can be configured with following options.

* `CORS_ORIGIN_ALLOW_ALL`: If True, the whitelist will not be used and all origins will be accepted. (default: False)
* `CORS_ORIGIN_WHITELIST`: A list of origin hostnames (including the scheme and with optional port) that are authorized to make cross-site HTTP requests.

## Historical API
If you wish to expose an API for querying previous revisions of documents, you need to set following environment variable:

`ENABLE_HISTORICAL_API`: Defaults to `false`.

If you enable this, make sure to also configure [visibilities](extending.md#visibility-classes) and
[permissions](extending.md#permission-classes) for historical types.

## Access Logging

The `caluma_logging` app bundles a middleware that logs all requests using the `AccessLog` model.
The log entries consist of username and caluma query/mutation details.

It's enabled by setting the environment variable:
`ENABLE_ACCESS_LOG`: Defaults to `false`.

For cleaning up old entries it's encouraged to use the `cleanup_access_log` management command.
Given the `--force` argument, it removes all entries older than what's given by `--keep` (default: "2 weeks").

## File question and answers
In order to make use of Calumas file question and answer, you need to set up a storage provider.

For the time being, only [MinIO](https://min.io/) is supported. Other providers may follow.

In the [docker-compose.yml](../docker-compose.yml)
you can find an example configuration for a MinIO container.

The following environment variables need to be set for caluma to use MinIO:

* `MEDIA_STORAGE_SERVICE`: defaults to "minio" (this is the only supported type
   at this time)
* `MINIO_STORAGE_ENDPOINT`: defaults to "minio:9000"
* `MINIO_STORAGE_ACCESS_KEY`: defaults to "minio"
* `MINIO_STORAGE_SECRET_KEY`: defaults to "minio123"
* `MINIO_STORAGE_USE_HTTPS`: defaults to False
* `MINIO_STORAGE_MEDIA_BUCKET_NAME`: defaults to "caluma-media"
* `MINIO_STORAGE_AUTO_CREATE_MEDIA_BUCKET`: defaults to True
* `MINIO_PRESIGNED_TTL_MINUTES`: defaults to 15

Caluma only handles metadata about files, not the files itself. When saving a `FileAnswer`, Caluma
will return a presigned `uploadUrl`, which the client can use to upload the file directly to the storage provider.

The same goes for retrieving files. Caluma will respond with a presigned `downloadUrl` for
the client to directly download the file from the storage provider.

In case you run HTTPS in your local development environment, you might have a
self-signed certificate. The Minio client by default verifies the TLS
certificates, so would fail in this case. You can set `MINIO_DISABLE_CERT_CHECKS`
to `true` to avoid this. Note: This setting only works if you also set `DEBUG`
to `true` as well.

## Client tokens
If you want to use additional services that need to talk to caluma (e.g.
[caluma-interval](https://github.com/projectcaluma/caluma-interval)), you need to have
an additional OIDC-client with the `token_introspection` scope key.

Following environment variables need to be set for caluma:

* `OIDC_INTROSPECT_ENDPOINT`: introspect endpoint from the OIDC-provider
* `OIDC_INTROSPECT_CLIENT_ID`: ID of the OIDC-client
* `OIDC_INTROSPECT_CLIENT_SECRET`: Secret of the OIDC-client

Some OIDC implementations (e.g. keycloak), allow for querying the `userinfo` endpoint
with a client token. In that case the `introspection` endpoint is never called.

The attribute `claims_source` on `OIDCUser` instances indicates the source of the claims.


## uWSGI defaults

We are using the sane uWSGI-defaults researched by [bloomberg](https://www.techatbloomberg.com/blog/configuring-uwsgi-production-deployment/?sf104898833=1). You can override the defaults using environment variables.

- `UWSGI_STRICT`=`true`
- `UWSGI_WSGI_FILE`=`/app/caluma/wsgi.py`
- `UWSGI_MASTER`=`true`
- `UWSGI_ENABLE_THREADS`=`true`
- `UWSGI_VACUUM`=`true`
- `UWSGI_SINGLE_INTERPRETER`=`true`
- `UWSGI_DIE_ON_TERM`=`true`
- `UWSGI_NEED_APP`=`true`
- `UWSGI_DISABLE_LOGGING`=`true`
- `UWSGI_LOG_4XX`=`true`
- `UWSGI_LOG_5XX`=`true`
- `UWSGI_MAX_REQUESTS`=`1000`
- `UWSGI_MAX_WORKER_LIFETIME`=`3600`
- `UWSGI_RELOAD_ON_RSS`=`2048`
- `UWSGI_WORKER_RELOAD_MERCY`=`60`
- `UWSGI_CHEAPER_ALGO`=`busyness`
- `UWSGI_PROCESSES`=`500`
- `UWSGI_CHEAPER`=`8`
- `UWSGI_CHEAPER_INITIAL`=`16`
- `UWSGI_CHEAPER_OVERLOAD`=`1`
- `UWSGI_CHEAPER_STEP`=`16`
- `UWSGI_CHEAPER_BUSYNESS_MULTIPLIER`=`30`
- `UWSGI_CHEAPER_BUSYNESS_MIN`=`20`
- `UWSGI_CHEAPER_BUSYNESS_MAX`=`70`
- `UWSGI_CHEAPER_BUSYNESS_BACKLOG_ALERT`=`16`
- `UWSGI_CHEAPER_BUSYNESS_BACKLOG_STEP`=`2`
- `UWSGI_HARAKIRI`=`60`
- `UWSGI_AUTO_PROCNAME`=`true`
- `UWSGI_PROCNAME_PREFIX`=`"caluma "`

See https://git.io/JemA2 for more information.


## Healthz Endpoint
The `healthz` endpoint performs a number of health checks using the
django-watchman package extended by custom checks. This allows systems, such
as Kubernetes, to perform monitoring and readiness checks on caluma.

Please note that when you enable it, make sure it is not reachable from the
outside (proxy configuration!). Due to it's nature, it takes some resources to
check everything, which makes it prone to abuse (denial of service, DoS).

It's enabled by setting the environment variable:
`ENABLE_HEALTHZ_ENDPOINT`: Defaults to `false`.


CHANGELOG.md
# v7.13.0

### Feature
* **config:** Introduce configurable user factory ([`8a49aaf`](https://github.com/projectcaluma/caluma/commit/8a49aaf7f911d4df50a354915530bfcd034ec91b))

### Fix
* **documentation:** Updated readme to reflect that GraphiQL is not available anymore ([`b6204e8`](https://github.com/projectcaluma/caluma/commit/b6204e8c89308a4878590c7026744af547f69a90))
* **serializers:** Do not implicitly monkeypatch rest framework ([`86b8a05`](https://github.com/projectcaluma/caluma/commit/86b8a05eaa7de391a2e650445f4f4cbb2ac02d25))
* Prettier choice question error message ([`acbc25c`](https://github.com/projectcaluma/caluma/commit/acbc25ca55b6e0dc339ec67ee19e1f712376938b))



# v7.12.0 (23 June 2021)

### Feature
* **jexl:** Allow optional answer transforms with default value ([`72dfa5b`](https://github.com/projectcaluma/caluma/commit/72dfa5bdf532025cb4ea35d671b944c011969a4a))


# v7.11.3 (10 June 2021)

Minor release only containing dependency updates. The release is done mainly because if this commit:

 * chore: update pyjexl [`03546ce`]\
   The bugfix https://github.com/mozilla/pyjexl/pull/23 (short-circuit
   evaluation) is critical for us, so we're pulling it in explicitly until
   there is a new pyjexl release.


# v7.11.2 (8 June 2021)

### Fix
* Extend uwsgi buffer ([`324d9d3`](https://github.com/projectcaluma/caluma/commit/324d9d397173cf2033600d12316dbce7bf23156c))
* **form:** Fetch selected_options of corresp. question and document ([`df97a5d`](https://github.com/projectcaluma/caluma/commit/df97a5de83c3190164dda57aa3712d5b421b4867))

# v7.11.1 (7 June 2021)

### Fix
* **validation:** Exclude hidden cells in answer transforms for tables ([`36b0972`](https://github.com/projectcaluma/caluma/commit/36b09720da6ef3e5a57da725ce65c1b6ba4e9ee3))


# v7.11.0 (28 May 2021)

### Feature
* Make selected option(s) accessible from answers
([`824d9a8`](https://github.com/projectcaluma/caluma/commit/824d9a87bf0ffb498e11b93388e84150c497ae2b))
* **forms:** Add question filter for forms
([`142e18b`](https://github.com/projectcaluma/caluma/commit/142e18b592a8d1ca6987be18f2245edbc4577ccd))
* **question:** Add filters for sub and row form on questions
([`3b14abf`](https://github.com/projectcaluma/caluma/commit/3b14abf825840aeaf9e97e0a2c3c4085a43f9bb0))


# v7.10.0 (23 April 2021)

### Feature
* Add IN lookup for json fields (#1472) (95bc2c3c88cab3d68137996a3daafdf3a1dd6c42)

### Fix
* Fix regression for union visibilities (e9a948c851e1bb9266e3e64205c9f21d0b3dca72)


# v7.9.1 (20 April 2021)

### Fix
* **document:** Pass current user to copy document function ([`178602f`](https://github.com/projectcaluma/caluma/commit/178602f1d57c1413cbc5d84a38f98923fa4c4f10))


# v7.9.0 (31 March 2021)

### Feature
* Health endpoint ([#1354](https://github.com/projectcaluma/caluma/issues/1354)) ([`bfc4951`](https://github.com/projectcaluma/caluma/commit/bfc49516caafaef216d00bfec4bdfb1047af77c9))

### Fix
* **workflow:** Correctly set workitem name from task ([#1448](https://github.com/projectcaluma/caluma/issues/1448)) ([`1095de4`](https://github.com/projectcaluma/caluma/commit/1095de43e319c4b89ab80a304ff031763beca4c4))


# v7.8.1 (30. March 2021)

### Fix
* **form:** Modified_content on row documents ([`75822ef`](https://github.com/projectcaluma/caluma/commit/75822efd29bc5cfc96af205631cfb4975666b43a))
* **file:** Ignore errors due to missing file ([#1440](https://github.com/projectcaluma/caluma/issues/1440)) ([`63d9f94`](https://github.com/projectcaluma/caluma/commit/63d9f94a5fb7de749dd2074f28acdaebd3409ec5))


# v7.8.0 (16. March 2021)

### Feature

* Improve mapby and add stringify transform ([`5fc1c73`](https://github.com/projectcaluma/caluma/commit/5fc1c737f15deab3e4ca7457dedac8f6ba00bd82))
* Modified_content properties on document ([`e333af0`](https://github.com/projectcaluma/caluma/commit/e333af0ca5fb251d00e8c65a1c4dfc3078c55dec))
* Add modified_by_user and modified_by_group fields ([`2868866`](https://github.com/projectcaluma/caluma/commit/286886689a9ca3908b9e84c96e0c3c54949a872f))
* Allow disabling ssl verification (#1408) ([`e0d6652`](https://github.com/projectcaluma/caluma/commit/e0d6652d1274ac8f8d8a6c2f901df82c2970b9c6))

### Fix
* Minio client requires different call signature for copy ([`f878bf1`](https://github.com/projectcaluma/caluma/commit/f878bf17d67dc819b44f0a40fca458cac7e6b7d1))
* Remove unneeded default_answer field on serializer ([`3104840`](https://github.com/projectcaluma/caluma/commit/3104840f9a9c7b48c203f2b36171cf3e2b208c2a))
* New minio has different metadata structure (#1412) ([`4f021af`](https://github.com/projectcaluma/caluma/commit/4f021af1fceb36ccedd7d0671b79c66fb50f1e0a))
* Update minio client to adjust for changed exceptions ([`0783629`](https://github.com/projectcaluma/caluma/commit/07836299823ac32c0591d8034a5f9f9d92321596))
* Use fixed maxsize from newest version ([`a0ac175`](https://github.com/projectcaluma/caluma/commit/a0ac175ce6674485030b82b2df46239e05663931))
* Fix search filter when used in collection (#1392) ([`d99db68`](https://github.com/projectcaluma/caluma/commit/d99db6858b9339704bd8b52fb2d2b8eb0dea4123))

# v7.7.1 (24. February 2021)

### Fix
* Extract_global_id should not be too eager in decoding
([#1389](https://github.com/projectcaluma/caluma/issues/1389))
([`8f97dc6`](https://github.com/projectcaluma/caluma/commit/8f97dc6ef30684653d1eba8036606d28386e554a))

# v7.7.0 (12. February 2021)

### Feature
* **form:** Add python API and command for copying forms ([`7620eec`](https://github.com/projectcaluma/caluma/commit/7620eec898b70dd641b758f0f4e09c6d3bdbacf8))
* **jexl:** Add min, max, sum, round transforms ([`beb714a`](https://github.com/projectcaluma/caluma/commit/beb714ad3c95ea4d32ef81ede84c023ed635511c))
* Dump_caluma command for entire forms ([#1367](https://github.com/projectcaluma/caluma/issues/1367)) ([`20b2a79`](https://github.com/projectcaluma/caluma/commit/20b2a79fea22f2b197d5f0238b33089ce063cde5))

### Fix

caluma/settings.py
import os
import re

import environ
from django.conf import global_settings
from promise import async_instance

env = environ.Env()
django_root = environ.Path(__file__) - 2

ENV_FILE = env.str("ENV_FILE", default=django_root(".env"))
if os.path.exists(ENV_FILE):  # pragma: no cover
    environ.Env.read_env(ENV_FILE)

# per default production is enabled for security reasons
# for development create .env file with ENV=development
ENV = env.str("ENV", "production")


def default(default_dev=env.NOTSET, default_prod=env.NOTSET):
    """Environment aware default."""
    return default_prod if ENV == "production" else default_dev


SECRET_KEY = env.str("SECRET_KEY", default=default("uuuuuuuuuu"))
DEBUG = env.bool("DEBUG", default=default(True, False))
ALLOWED_HOSTS = env.list("ALLOWED_HOSTS", default=default(["*"]))


# Application definition

INSTALLED_APPS = [
    "django.contrib.postgres",
    "localized_fields",
    "psqlextra",
    "django.contrib.contenttypes",
    "django.contrib.staticfiles",
    "graphene_django",
    "simple_history",
    "caluma.caluma_core.apps.DefaultConfig",
    "caluma.caluma_user.apps.DefaultConfig",
    "caluma.caluma_form.apps.DefaultConfig",
    "caluma.caluma_workflow.apps.DefaultConfig",
    "caluma.caluma_data_source.apps.DefaultConfig",
]

if DEBUG:
    INSTALLED_APPS.append("django_extensions")

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.locale.LocaleMiddleware",
    "simple_history.middleware.HistoryRequestMiddleware",
]

ROOT_URLCONF = "caluma.urls"
WSGI_APPLICATION = "caluma.wsgi.application"
TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [django_root("caluma", "templates")],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.messages.context_processors.messages",
            ]
        },
    }
]


# Database
# https://docs.djangoproject.com/en/1.11/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": "psqlextra.backend",
        "NAME": env.str("DATABASE_NAME", default="caluma"),
        "USER": env.str("DATABASE_USER", default="caluma"),
        "PASSWORD": env.str("DATABASE_PASSWORD", default=default("caluma")),
        "HOST": env.str("DATABASE_HOST", default="localhost"),
        "PORT": env.str("DATABASE_PORT", default=""),
    }
}

# Cache
# https://docs.djangoproject.com/en/1.11/ref/settings/#caches

CACHES = {
    "default": {
        "BACKEND": env.str(
            "CACHE_BACKEND", default="django.core.cache.backends.locmem.LocMemCache"
        ),
        "LOCATION": env.str("CACHE_LOCATION", ""),
    }
}

# Internationalization
# https://docs.djangoproject.com/en/1.11/topics/i18n/


def parse_languages(languages):
    return [(language, language) for language in languages]


LANGUAGE_CODE = env.str("LANGUAGE_CODE", "en")
LANGUAGES = (
    parse_languages(env.list("LANGUAGES", default=[])) or global_settings.LANGUAGES
)

TIME_ZONE = env.str("TIME_ZONE", "UTC")
USE_I18N = True
USE_L10N = True
USE_TZ = True


# Managing files

MEDIA_STORAGE_SERVICE = env.str("MEDIA_STORAGE_SERVICE", default="minio")
MINIO_STORAGE_ENDPOINT = env.str("MINIO_STORAGE_ENDPOINT", default="minio:9000")
MINIO_STORAGE_ACCESS_KEY = env.str("MINIO_STORAGE_ACCESS_KEY", default="minio")
MINIO_STORAGE_SECRET_KEY = env.str("MINIO_STORAGE_SECRET_KEY", default="minio123")
MINIO_STORAGE_USE_HTTPS = env.str("MINIO_STORAGE_USE_HTTPS", default=False)
MINIO_STORAGE_MEDIA_BUCKET_NAME = env.str(
    "MINIO_STORAGE_MEDIA_BUCKET_NAME", default="caluma-media"
)
MINIO_STORAGE_AUTO_CREATE_MEDIA_BUCKET = env.str(
    "MINIO_STORAGE_AUTO_CREATE_MEDIA_BUCKET", default=True
)
MINIO_PRESIGNED_TTL_MINUTES = env.str("MINIO_PRESIGNED_TTL_MINUTES", default=15)


def parse_admins(admins):  # pragma: no cover
    """
    Parse env admins to django admins.

    Example of ADMINS environment variable:
    Test Example <test@example.com>,Test2 <test2@example.com>
    """
    result = []
    for admin in admins:
        match = re.search(r"(.+) \<(.+@.+)\>", admin)
        if not match:
            raise environ.ImproperlyConfigured(
                'In ADMINS admin "{0}" is not in correct '
                '"Firstname Lastname <email@example.com>"'.format(admin)
            )
        result.append((match.group(1), match.group(2)))
    return result


ADMINS = parse_admins(env.list("ADMINS", default=[]))


# GraphQL

GRAPHENE = {"SCHEMA": "caluma.schema.schema", "MIDDLEWARE": []}

if DEBUG:
    GRAPHENE["MIDDLEWARE"].append("graphene_django.debug.DjangoDebugMiddleware")

# OpenID connect

OIDC_USERINFO_ENDPOINT = env.str("OIDC_USERINFO_ENDPOINT", default=None)
OIDC_VERIFY_SSL = env.bool("OIDC_VERIFY_SSL", default=True)
OIDC_GROUPS_CLAIM = env.str("OIDC_GROUPS_CLAIM", default="caluma_groups")
OIDC_USERNAME_CLAIM = env.str("OIDC_USERNAME_CLAIM", default="sub")
OIDC_BEARER_TOKEN_REVALIDATION_TIME = env.int(
    "OIDC_BEARER_TOKEN_REVALIDATION_TIME", default=0
)

OIDC_INTROSPECT_ENDPOINT = env.str("OIDC_INTROSPECT_ENDPOINT", default=None)
OIDC_INTROSPECT_CLIENT_ID = env.str("OIDC_INTROSPECT_CLIENT_ID", default=None)
OIDC_INTROSPECT_CLIENT_SECRET = env.str("OIDC_INTROSPECT_CLIENT_SECRET", default=None)

# Extensions

VISIBILITY_CLASSES = env.list(
    "VISIBILITY_CLASSES", default=default(["caluma.caluma_core.visibilities.Any"])
)

PERMISSION_CLASSES = env.list(
    "PERMISSION_CLASSES", default=default(["caluma.caluma_core.permissions.AllowAny"])
)

VALIDATION_CLASSES = env.list("VALIDATION_CLASSES", default=[])

DATA_SOURCE_CLASSES = env.list("DATA_SOURCE_CLASSES", default=[])

FORMAT_VALIDATOR_CLASSES = env.list("FORMAT_VALIDATOR_CLASSES", default=[])

# Cors headers
CORS_ORIGIN_ALLOW_ALL = env.bool("CORS_ORIGIN_ALLOW_ALL", default=False)
CORS_ORIGIN_WHITELIST = env.list("CORS_ORIGIN_WHITELIST", default=[])

# simple history
SIMPLE_HISTORY_HISTORY_ID_USE_UUID = True

# Historical API
ENABLE_HISTORICAL_API = env.bool("ENABLE_HISTORICAL_API", default=False)

# Logging

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "handlers": {"console": {"class": "logging.StreamHandler"}},
    "loggers": {
        "": {"handlers": ["console"], "level": env.str("LOG_LEVEL", default="INFO")}
    },
}

# Configure the fields you intend to use in the "meta" fields. This will
# provide corresponding constants in the ordreBy filter.
META_FIELDS = env.list("META_FIELDS", default=[])


# static files

STATIC_URL = "/static/"

# disable trampoline to improve performance
# see https://github.com/graphql-python/graphene/issues/268
# this means though that data loader won't work which we currently don't use
# best to remove once graphql-core-next is compatible with graphene.
async_instance.disable_trampoline()
